### PR TITLE
Added Media Selection

### PR DESCRIPTION
**Pull request by 3rfan**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

The media selection content type.

#### Why?

A feature is needed to assign media to a form.

### CODE REVIEW

**Comment by 3rfan** [comment]
@danrot

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+            this.clearSelection = jest.fn();
+            this.setAppendRequestData = jest.fn();
+            this.deselectEntirePage = jest.fn();
+            this.getSchema = jest.fn().mockReturnValue({});
+        }),
+    };
+});
+
+jest.mock('sulu-admin-bundle/containers/Datagrid/registries/DatagridAdapterRegistry', () => {
+    const getAllAdaptersMock = jest.fn();
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: This const is probably not really necessary, just inline the `jest.fn()` in the `return`.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+jest.mock('sulu-admin-bundle/services', () => ({
+    translate: function(key) {
+        switch (key) {
+            case 'sulu_media.all_media':
+                return 'All Media';
+            case 'sulu_media.copy_url':
+                return 'Copy URL';
+            case 'sulu_media.download_masterfile':
+                return 'Downoad master file';
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: `Download`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+            case 'sulu_admin.object':
+                return 'Object';
+            case 'sulu_admin.objects':
+                return 'Objects';
+        }
+    },
+}));
+
+beforeEach(() => {
+    const datagridAdapterRegistry = require('sulu-admin-bundle/containers/Datagrid/registries/DatagridAdapterRegistry');
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: Any good reason not to `import` the file on top instead?

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+                return 'of';
+            case 'sulu_admin.object':
+                return 'Object';
+            case 'sulu_admin.objects':
+                return 'Objects';
+        }
+    },
+}));
+
+beforeEach(() => {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: I would not place the `beforeEach` in the middle of multiple `jest.mock`, but move it after them instead.

File: src/Sulu/Bundle/MediaBundle/Resources/translations/admin.de.json [hunk 5]
```diff
@@ -1,5 +1,7 @@
 {
     "sulu_media.all_media": "Alle Medien",
     "sulu_media.copy_url": "URL kopieren",
-    "sulu_media.download_masterfile": "Download master file"
+    "sulu_media.download_masterfile": "Download master file",
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/translations/admin.de.json]: Das ist eigentlich nicht wirklich Englisch :D

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            options,
+            true,
+            this.handleMediaSelection,
+            this.selectedMediaIds
+        );
+    }
+
+    addMediaToSelectionStore() {
+        this.selectedMediaIds.forEach((mediaId) => {
+            const media = this.mediaDatagridStore.data.find((item) => item.id === mediaId);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js]: I think that can't really work... What if you've added a media, change the collection afterwards, and press ok then? Then the store you are accessing here wouldn't have the item you are looking for anymore.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+    destroyOverlay() {
+        this.selectedMediaIds = [];
+        this.collectionId = undefined;
+
+        this.destroyStores();
+        this.overlayDisposer();
+    }
+
+    createStores = () => {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js]: I think also all of this stuff could go into the separate `MediaSelectionOverlay` component. Would make this file a lot cleaner IMO.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Backdrop/backdrop.scss [hunk 8]
```diff
@@ -6,6 +6,7 @@ $backdropBackground: rgba($dustyGray, .9);
     position: fixed;
     top: 0;
     left: 0;
+    z-index: 100;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Backdrop/backdrop.scss]: The toolbar because of the added `z-index`. That was added to fix an layering issue with the masonry view.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/MultiItemSelection/Item.js [hunk 9]
```diff
@@ -47,6 +47,7 @@ export default class Item extends React.PureComponent<Props> {
                 </div>
                 {onRemove &&
                     <button
+                        type="button"
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/MultiItemSelection/Item.js]: yes

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/Overlay.js [hunk 10]
```diff
@@ -97,33 +97,35 @@ export default class Overlay extends React.PureComponent<Props> {
...
+                                <footer>
+                                    <Actions actions={actions} />
+                                    <Button skin="primary" onClick={onConfirm}>
+                                        {confirmText}
+                                    </Button>
+                                </footer>
+                            </section>
+                        </div>
                     </div>
-                    <Backdrop local={true} onClick={onClose} />
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/Overlay.js]: No, the `local` param is also used in the toolbar.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/overlay.scss [hunk 11]
```diff
@@ -21,6 +21,7 @@ $transitionDuration: 300ms;
     position: absolute;
     top: 0;
     left: 0;
+    z-index: 100;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/overlay.scss]: Answer above.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/tests/__snapshots__/Overlay.test.js.snap [hunk 12]
```diff
@@ -1,11 +1,14 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`The component should not render in body when closed 1`] = `null`;
+exports[`The component should not render in body when closed 1`] = `<div />`;
 
-exports[`The component should render in body when open 1`] = `null`;
+exports[`The component should render in body when open 1`] = `<div />`;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/tests/__snapshots__/Overlay.test.js.snap]: That is not possible because the outputted `div` is a wrapper to enclose multiple `jsx` tags.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Popover/popover.scss [hunk 13]
```diff
@@ -5,6 +5,6 @@
     left: 0;
     width: 100%;
     height: 100%;
-    z-index: 1;
+    z-index: 100;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Popover/popover.scss]: Same as above

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/OptionList.js [hunk 14]
```diff
@@ -45,6 +45,7 @@ export default class OptionList extends React.PureComponent<Props> {
 
         return (
             <div>
+                <Backdrop open={true} local={true} onClick={this.handleBackdropClick} visible={false} />
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/OptionList.js]: It is `local` because the `OptionList` is inside the `Toolbar` and not a `Popover` kind of component. We need the `local` param to achieve the needed behavior.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/optionList.scss [hunk 15]
```diff
@@ -8,7 +8,7 @@ $optionListShadowColor: $silver;
     position: absolute;
     left: 0;
     top: 100%;
-    z-index: 1;
+    z-index: 100;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/optionList.scss]: Same as above

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/toolbar.scss [hunk 16]
```diff
@@ -5,6 +5,8 @@
     margin: 0;
     background: $toolbarBackgroundColor;
     box-shadow: 2px 6px 12px 0 rgba($shadowColor, .5);
+    position: relative;
+    z-index: 1;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/toolbar.scss]: To fix that mentioned masonry issue

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    prepareMedia(media: Object) {
+        return {
+            id: media.id,
+            title: media.title,
+            thumbnail: media.thumbnails[THUMBNAIL_SIZE],
+        };
+    }
+
+    loadSelectedMedia = (selectedMediaIds: Array<string | number>, locale: string) => {
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: yes

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js (hunk 1 above)
Comment (3rfan) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: It is for line `85`.

**Comment by danrot** [comment]
Without having a look at the code, I got an answer from the creator:
image

I really like the idea with a computed getter, that would probably be the way to go with the problem we were experiencing today.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    componentWillMount() {
+        const {
+            open,
+            locale,
+        } = this.props;
+
+        if (open) {
+            this.overlayDisposer = autorun(() => this.createStores(locale));
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: That's where we should probably use the `computed` suggestion from Michel.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    componentWillReceiveProps(nextProps: Props) {
+        const {
+            open,
+            locale,
+        } = this.props;
+
+        if (!open && nextProps.open) {
+            this.overlayDisposer = autorun(() => this.createStores(locale));
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: Same as above

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+@observer
+export default class MediaSelectionOverlay extends React.PureComponent<Props> {
+    static defaultProps = {
+        open: false,
+        excludedIds: [],
+    };
+
+    mediaPage: observable = observable();
+    collectionPage: observable = observable();
+    @observable collectionId: ?string | number;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: This could be just `number`, right?

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+
+    @action setMediaPage(page: number) {
+        this.mediaPage.set(page);
+    }
+
+    @action setCollectionPage(page: number) {
+        this.collectionPage.set(page);
+    }
+
+    @action setCollectionId(id: ?string | number) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: `id` could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+        if (this.mediaDatagridStore) {
+            this.mediaDatagridStore.destroy();
+        }
+
+        if (this.collectionDatagridStore) {
+            this.collectionDatagridStore.destroy();
+        }
+    }
+
+    @action createCollectionDatagridStore(collectionId: ?string | number, page: observable, locale: string) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: `collectionId` could be just number

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+            COLLECTIONS_RESOURCE_KEY,
+            {
+                page,
+                locale,
+            },
+            (collectionId) ? {parent: collectionId} : undefined
+        );
+    }
+
+    createCollectionStore = (collectionId: ?string | number, locale: string) => {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+        if (this.collectionStore) {
+            this.collectionStore.destroy();
+        }
+
+        this.collectionStore = new CollectionStore(collectionId, locale);
+    };
+
+    @action createMediaDatagridStore(collectionId: ?string | number, page: observable, locale: string) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: Just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+        ].join(',');
+
+        page.set(1);
+
+        if (collectionId) {
+            options.collection = collectionId;
+        }
+
+        if (excludedIds.length) {
+            options['excluded-ids'] = excludedIds.join(',');
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: We've decided to go for `excluded` here, because we have that already in other requests, and we would like to stay consistent.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+            },
+            options,
+            true,
+            this.handleMediaSelectionChange
+        );
+
+        this.selectedMedia.forEach((media) => this.mediaDatagridStore.select(media.id));
+    }
+
+    handleMediaSelectionChange = (mediaId: string | number, selected: boolean) => {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+
+            if (media) {
+                this.selectedMedia.push(media);
+            }
+        } else {
+            this.selectedMedia = this.selectedMedia.filter((media) => media.id !== mediaId);
+        }
+    };
+
+    handleCollectionNavigate = (collectionId: ?string | number) => {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelectionOverlay.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+import MediaSelectionStore from './stores/MediaSelectionStore';
+import MediaSelectionOverlay from './MediaSelectionOverlay';
+import MediaSelectionItem from './MediaSelectionItem';
+
+const ADD_ICON = 'plus';
+
+type Props = {
+    locale: observable,
+    value: Array<string | number>,
+    onChange: (ids: Array<string | number>) => void,
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js]: Both of the above could just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+import {ResourceRequester} from 'sulu-admin-bundle/services';
+import type {MediaItem} from '../types';
+
+const THUMBNAIL_SIZE = 'sulu-25x25';
+const MEDIA_RESOURCE_KEY = 'media';
+
+export default class MediaSelectionStore {
+    @observable selectedMedia: Array<MediaItem> = [];
+
+    constructor(selectedMediaIds: ?Array<string | number>, locale: string) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+            this.loadSelectedMedia(selectedMediaIds, locale);
+        }
+    }
+
+    @action add(media: Object) {
+        const preparedMedia = this.prepareMedia(media);
+        this.selectedMedia.push(preparedMedia);
+    }
+
+    @action removeById(mediaId: string | number) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+        this.selectedMedia = this.selectedMedia.filter((media) => media.id !== mediaId);
+    }
+
+    @action move(oldItemIndex: number, newItemIndex: number) {
+        this.selectedMedia = arrayMove(this.selectedMedia, oldItemIndex, newItemIndex);
+    }
+
+    @computed get selectedMediaIds(): Array<string | number> {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: Could be just `number`

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: And it could be just `number` here 🙈

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js [hunk 32]
```diff
@@ -0,0 +1,294 @@
...
+import datagridAdapterRegistry from 'sulu-admin-bundle/containers/Datagrid/registries/DatagridAdapterRegistry';
+import MediaSelection from '../MediaSelection';
+import MediaCardSelectionAdapter from '../../Datagrid/adapters/MediaCardSelectionAdapter';
+import MediaSelectionStore from '../stores/MediaSelectionStore';
+
+jest.mock('../stores/MediaSelectionStore', () => jest.fn());
+
+jest.mock('sulu-admin-bundle/containers', () => {
+    return {
+        Datagrid: require('sulu-admin-bundle/containers/Datagrid/Datagrid').default,
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: Was there now any reason not to import that with the `import` above?

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/types.js [hunk 33]
```diff
@@ -0,0 +1,6 @@
+// @flow
+export type MediaItem = {
+    id: string | number,
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/types.js]: Could be just `number` 🙈

File: src/Sulu/Bundle/MediaBundle/Resources/js/stores/CollectionStore/CollectionStore.js [hunk 34]
```diff
@@ -5,15 +5,15 @@ import type {BreadcrumbItem, BreadcrumbItems, Collection} from './types';
...
+export default class CollectionStore {
     @observable loading: boolean = false;
     @observable collection: Collection = {
         parentId: null,
         breadcrumb: null,
     };
     disposer: () => void;
 
-    constructor(collectionId: ?number, locale: observable) {
+    constructor(collectionId: ?string | number, locale: observable) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/stores/CollectionStore/CollectionStore.js]: Why did you add the `string` type here?

File: src/Sulu/Bundle/MediaBundle/Resources/js/stores/CollectionStore/CollectionStore.js [hunk 35]
```diff
@@ -35,7 +35,7 @@ export default class CollectionInfoStore {
         this.loading = loading;
     }
 
-    @action load(collectionId: ?number, locale: string) {
+    @action load(collectionId: ?string | number, locale: string) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/stores/CollectionStore/CollectionStore.js]: Same as above

**Comment by danrot** [comment]
This  `string|number` thing really bothers me... Please check all the occurences once more. Having both options at a place where only on can occur means that there is more complexity than needed.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/Overlay.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/Overlay.js]: The `Backdrop` being local and inside the `Portal` of the `Overlay` leaded to weird layering behavior after the `z-index` change which could have been fixed by giving the `Backdrop` a higher `z-index`.  I thought changing the `jsx` structure is the better approach.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/toolbar.scss (hunk 16 above)
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Toolbar/toolbar.scss]: Did we? So we go with the approach to add a container around the `Toolbar`? The current structure looks like the following:

~~~javascript
<div>
    <Toolbar onNavigationButtonClick={this.handleNavigationButtonClick} />
    <main className={applicationStyles.main}>
        {router.route &&
            <ViewRenderer
                key={router.route.name}
                router={router}
            />
        }
    </main>
    <SplitView />
    <Overlay
        title="Navigation"
...
~~~

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/mediaSelectionItem.scss [hunk 36]
```diff
@@ -0,0 +1,17 @@
...
+    align-items: center;
+}
+
+.thumbnail {
+    margin: 0 20px 0 10px;
+
+    img {
+        display: block;
+        width: $thumbnailSize;
+        height: $thumbnailSize;
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/mediaSelectionItem.scss]: It is needed for when the image is not loaded yet

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js (hunk 32 above)
Comment (3rfan) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/tests/MediaSelection.test.js]: The first comment was referring to the 'datagridAdapterRegistry' which is not included through `import`. I have to load the `Datagrid` this way because it is loaded through `named imports` inside the code in the `sulu-media-bundle`

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Field.js [hunk 37]
```diff
@@ -19,14 +20,22 @@ export default class Field extends React.PureComponent<Props> {
     };
 
     render() {
-        const {schema, value} = this.props;
+        const {
+            value,
+            locale,
+            schema,
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Field.js]: Wouldn't do that in this PR and would be nice if we could add an `eslint` rule for that

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/Datagrid/adapters/MediaCardAdapter.js [hunk 38]
```diff
@@ -73,9 +76,10 @@ export default class MediaCardAdapter extends React.Component<Props> {
                             icon={icon}
                             title={item.title}
                             image={item.thumbnails[THUMBNAIL_SIZE]}
-                            onClick={onItemClick}
-                            selected={selections.includes(item.id)}
-                            onSelectionChange={onItemSelectionChange}
+                            onClick={onItemSelectionChange}
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/Datagrid/adapters/MediaCardAdapter.js]: ups

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Field.js (hunk 37 above)
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Field.js]: You have literally introduced that change in this PR 🙈

[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Field.js]: And I have already been looking for a rule, but couldn't find one 😕

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Form.js [hunk 39]
```diff
@@ -55,6 +55,7 @@ export default class Form extends React.PureComponent<Props> {
                 onChange={this.handleChange}
                 schema={schema}
                 data={store.data}
+                locale={store.locale}
```
[the remark this answers is not in the document]
Comment (3rfan) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/Form.js]: The `CollectionStore` which is also used by the `MediaOverview` needs the `locale` as an observable to update the breadcrumb on language change.

**Comment by 3rfan** [comment]
@danrot new day, new try!

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js (hunk 29 above)
Comment (danrot) [reply src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/stores/MediaSelectionStore.js]: Doesn't really make sense to have the optional parameter first, does it?

**Comment by 3rfan** [comment]
@danrot I will never give up! 
Alt Text

**Comment by danrot** [comment]
I have just tested the content on the website, and realized that it is not loading the media, but you retrieve a simple array instead. I guess this is because you simply send an array of IDs to the server, but the format should be like the following:

```javascript
{
    "displayOption": "top",
    "ids": [35,34]
}
```

I think just omitting the `displayOption` should be ok, as long as we are not implementing it. But the returned value from the `MediaSelection` should probably look something like this for now:

```javascript
{
    "ids": [35, 34]
}
```

I've used the [`sulu_snippet_load`](http://docs.sulu.io/en/latest/reference/twig-extensions/functions/sulu_snippet_load.html?highlight=sulu_snippet_load) to test that.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js [hunk 40]
```diff
@@ -12,8 +12,8 @@ const ADD_ICON = 'plus';
 
 type Props = {
     locale: observable,
-    value: ?Array<string | number>,
-    onChange: (ids: Array<string | number>) => void,
+    value: ?{ids: Array<string | number>},
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js]: I think we should move that into a separate type, shouldn't we? And that would go into a separate `types.js` file then.

File: src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js [hunk 41]
```diff
@@ -26,8 +26,13 @@ export default class MediaSelection extends React.PureComponent<Props> {
             value,
             locale,
         } = this.props;
+        let selectedMediaIds = null;
 
-        this.mediaSelectionStore = new MediaSelectionStore(value, locale);
+        if (value && value.ids) {
+            selectedMediaIds = value.ids;
+        }
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/js/containers/MediaSelection/MediaSelection.js]: How about `const selectedMediaIds = (value && value.ids) ? value.ids : null;`? Then it couldn't be changed by accident afterwards.

**Comment by 3rfan** [comment]
@danrot can we merge this...
Alt Text
